Support PEP-415's Exception.__suppress_context__

- [x] Include a detailed description of the bug or suggestion

PEP-415 states that `exception.__context__` should be suppressed in traceback outputs, if `exception.__suppress_context__` is `True`.

If a ``raise exception from None`` is caught by pytest, pytest should not chain the context in the test report.

The current algorithm in `_pytest._code.code.FormattedExcinfo` is:

```python
if e.__cause__ is not None:
    # Code to chain the cause.
elif e.__context__ is not None:
    # Code to chain the context.
```

which means that pytest always chains the exception, assuming that `e.__context__` hasn't been set to `None`.

By comparison, the algorithm in `traceback.TracebackException` is:

```python
if e.__cause__ is not None:
    # Code to chain the cause.
elif (e.__context__ is not None and not e.__suppress_context__):
    # Code to chain the context.
```

`Exception.__suppress_context__` is available in all of the versions of Python 3 that are supported by pytest, so it is trivial to add support for this feature.

- [x] Minimal example if possible

Here's a test that has a ``raise exception from None`` in it:

```python
def test_raise_from_none():
    try:
        raise ValueError()
    except Exception:
        raise AttributeError() from None
```

This is the test output (with the chained exception traceback) that results when running against the pytest feature branch (commit 768edde899fe3629a69d55289f82bb0d95635c06):

```
_____________________________________________________ test_raise_from_none _____________________________________________________

    def test_raise_from_none():
        try:
>           raise ValueError()
E           ValueError

testing/code/test_excinfo.py:1257: ValueError

During handling of the above exception, another exception occurred:

    def test_raise_from_none():
        try:
            raise ValueError()
        except Exception:
>           raise AttributeError() from None
E           AttributeError

testing/code/test_excinfo.py:1259: AttributeError
```

But running the same code in a terminal produces this shorter, non-chained traceback:

```
AttributeError                            Traceback (most recent call last)
<ipython-input-51-e2e3809c49fb> in <module>()
----> 1 test_raise_from_none()

<ipython-input-50-d651befdf00e> in test_raise_from_none()
      3         raise ValueError()
      4     except Exception:
----> 5         raise AttributeError() from None
      6

AttributeError:
```

And this is what the pytest output should look like:

```
_____________________________________________________ test_raise_from_none _____________________________________________________

    def test_raise_from_none():
        try:
            raise ValueError()
        except Exception:
>           raise AttributeError() from None
E           AttributeError

testing/code/test_excinfo.py:1259: AttributeError
```

--- a/AUTHORS
+++ b/AUTHORS
@@ -84,6 +84,7 @@ John Towler
 Jon Sonesen
 Jonas Obrist
 Jordan Guymon
+Jordan Moldow
 Joshua Bronson
 Jurko Gospodnetić
 Justyna Janczyszyn

--- a/_pytest/_code/code.py
+++ b/_pytest/_code/code.py
@@ -679,7 +679,7 @@ class FormattedExcinfo(object):
                     e = e.__cause__
                     excinfo = ExceptionInfo((type(e), e, e.__traceback__)) if e.__traceback__ else None
                     descr = 'The above exception was the direct cause of the following exception:'
-                elif e.__context__ is not None:
+                elif (e.__context__ is not None and not e.__suppress_context__):
                     e = e.__context__
                     excinfo = ExceptionInfo((type(e), e, e.__traceback__)) if e.__traceback__ else None
                     descr = 'During handling of the above exception, another exception occurred:'

--- a/changelog/2631.feature
+++ b/changelog/2631.feature
@@ -1,0 +1,4 @@
+Added support for `PEP-415's <https://www.python.org/dev/peps/pep-0415/>`_
+``Exception.__suppress_context__``. Now if a ``raise exception from None`` is
+caught by pytest, pytest will no longer chain the context in the test report.
+The behavior now matches Python's traceback behavior.

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -1096,6 +1096,36 @@ raise ValueError()
         assert tw.lines[47] == ":15: AttributeError"
 
     @pytest.mark.skipif("sys.version_info[0] < 3")
+    def test_exc_repr_with_raise_from_none_chain_suppression(self, importasmod):
+        mod = importasmod("""
+            def f():
+                try:
+                    g()
+                except Exception:
+                    raise AttributeError() from None
+            def g():
+                raise ValueError()
+        """)
+        excinfo = pytest.raises(AttributeError, mod.f)
+        r = excinfo.getrepr(style="long")
+        tw = TWMock()
+        r.toterminal(tw)
+        for line in tw.lines:
+            print(line)
+        assert tw.lines[0] == ""
+        assert tw.lines[1] == "    def f():"
+        assert tw.lines[2] == "        try:"
+        assert tw.lines[3] == "            g()"
+        assert tw.lines[4] == "        except Exception:"
+        assert tw.lines[5] == ">           raise AttributeError() from None"
+        assert tw.lines[6] == "E           AttributeError"
+        assert tw.lines[7] == ""
+        line = tw.get_write_msg(8)
+        assert line.endswith('mod.py')
+        assert tw.lines[9] == ":6: AttributeError"
+        assert len(tw.lines) == 10
+
+    @pytest.mark.skipif("sys.version_info[0] < 3")
     @pytest.mark.parametrize('reason, description', [
         ('cause', 'The above exception was the direct cause of the following exception:'),
         ('context', 'During handling of the above exception, another exception occurred:'),

